src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,358 @@
+        private ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>> _map;
+        // When the current cache size gets to this percentage of _capacity, _compactionPercentage% of the cache will be removed.
+        private readonly double _maxCapacityPercentage = .95;
+        private bool _disposed = false;
+        private readonly int _tryTakeTimeout;
+        // if true, expired values will not be added to the cache and clean-up of expired values will occur on a 5 minute interval
+        private readonly bool _removeExpiredValues;
+
+        internal EventBasedLRUCache(
+            int capacity,
+            TaskCreationOptions options = TaskCreationOptions.LongRunning,
+            IEqualityComparer<TKey> comparer = null,
+            int tryTakeTimeout = 500,
+            bool removeExpiredValues = true,
+            int cleanUpIntervalInSeconds = 300)
+        {
+            _tryTakeTimeout = tryTakeTimeout;
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _removeExpiredValues = removeExpiredValues;
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            if (_removeExpiredValues)
+                _ = RemoveExpiredValuesPeriodically(TimeSpan.FromSeconds(cleanUpIntervalInSeconds));
+        }
+
+        private void OnStart()
+        {
+            while (!_disposed)
+            {
+                try
+                {
+                    if (_eventQueue.TryTake(out var action, _tryTakeTimeout))
+                        action.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10900, ex));
+                }
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (!_disposed)
+            {
+                if (_eventQueue.Count == 0)
+                    return;
+            }
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            try
+            {
+                var node = _doubleLinkedList.First;
+                while (node != null)
+                {
+                    var nextNode = node.Next;
+                    if (node.Value.ExpirationTime < DateTime.UtcNow)
+                    {
+                        _doubleLinkedList.Remove(node);
+                        _map.TryRemove(node.Value.Key, out _);
```

Comment:
When a item is removed, when is it disposed?

Reply:
Either by the GC or by the owner, we didn't create and we don't know if another reference exists elsewhere.

Reply:
For SignatureProviders, TryRemove will zero out the cache reference, which CryptoProviderFactory.ReleaseSignatureProviderwill check if the ref is null null, when null, will dispose

Reply:
hmm, it doesn't seem right that the cache should control this life time. Seems like a responsibility of the CryptoProviderFactory or ideally the GC imo. Why would anyone use this object still if we have removed it?

---
In reply to: [592695435](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1605#discussion_r592695435) [](ancestors = 592695435)